Added Support For beginRequest and endRequest (new version)

fmeheust opened the pull request:
Continued work from pull request https://github.com/brettwooldridge/HikariCP/pull/2079. Addressed review comments https://github.com/brettwooldridge/HikariCP/pull/2079#discussion_r1367934838 and added tests.

Deeptanshu-Chowdhuri committed:
Done begin and endRequest.

Deeptanshu-Chowdhuri committed:
BeginRequest Changed a bit.

Deeptanshu-Chowdhuri committed:
Pool Entry changed only

Deeptanshu-Chowdhuri committed:
Removed Useless Comments

Deeptanshu-Chowdhuri committed:
Added Try catch to avoid changing function signature and stop at begin or endRequest.

Deeptanshu-Chowdhuri committed:
changed Log warning for begin and endRequest

Deeptanshu-Chowdhuri committed:
changed Formatting and commented changes.

Deeptanshu-Chowdhuri committed:
corrected DriverDataSource.java

Deeptanshu-Chowdhuri committed:
Removed comments to comply with the repository.

fmeheust committed:
Moved get JDBCVersion

fmeheust committed:
PoolEntry begin end request test

fmeheust committed:
typo

akashchouhan16 commented on src/main/java/com/zaxxer/hikari/pool/PoolEntry.java:
```diff
@@ -56,6 +56,7 @@ final class PoolEntry implements IConcurrentBagEntry
 
    private final boolean isReadOnly;
    private final boolean isAutoCommit;
+   private boolean isJDBC43OrLater;
```
The choice of variable is inconsistent with the codebase

akashchouhan16 commented on src/main/java/com/zaxxer/hikari/pool/PoolEntry.java:
```diff
@@ -70,6 +71,17 @@ final class PoolEntry implements IConcurrentBagEntry
       this.isAutoCommit = isAutoCommit;
       this.lastAccessed = currentTime();
       this.openStatements = new FastList<>(Statement.class, 16);
+      boolean isJDBC43OrLater = false;
```
any specific reason to use `isJDBC43OrLater` inside the static initializer block? Could this be moved?

lfbayer replied:
I imagine there is a desired to make the member variable `final` by having only a single line actually set the value. But this code still has a non-final member.

lfbayer commented on src/main/java/com/zaxxer/hikari/pool/PoolEntry.java:
```diff
@@ -69,6 +71,17 @@ final class PoolEntry implements IConcurrentBagEntry
+      boolean isJDBC43OrLater = false;
+      try {
+         DatabaseMetaData dm = connection.getMetaData();
+         isJDBC43OrLater = ((dm != null) && (
```
If the connection instance has the methods "beginRequest" and "endRequest", wouldn't that be sufficient to know we can call them, rather than doing a version check?

Maybe using reflection on the connection instance?

I'm wondering if adding the call to getMetaData has any performance implications.

It also seems like it would better to set this pool-wide, rather than have the value calculated for every new connection in the pool. I can't imagine that you would have a pool that would have different JDBC versions for each connection.

fmeheust replied:
Thank you for the feedback. I chose to do it for each connection because I thought that one could change the version of the driver that is in the application classpath and the pool would start using a new driver, but if that is not possible I propose the following approach:

I will move the check to PoolBase.checkDriverSupport(Connection) and store the value(s) in a private member. This/these value(s) will be added as a parameter on the constructor of the PoolEntry class, and passed in the PoolBase.newEntry() method.

Concerning using relection instead of database metadata, I could argue that if Hikari starts using new JDBC functionality in the future, the code could end up with several boolean variables (one for each method present in from specific version of the JDBC standard) instead of to int values jdbcMajorVersion and jdbcMinorVersion. I am new to Hikari, so please let me know which choice you prefer and I will comply.

lfbayer reviewed (changes requested):
It feels like moving this code to HikariPool getConnection an recycle might be a little better.

I also wonder if we need to make sure there is a way to explicitly disable having the pool call the begin and end methods.

fmeheust committed:
Moved request boundaries to HikariPool + property

fmeheust committed:
Corrected test

fmeheust committed:
Renamed property and tests

fmeheust committed:
Removed changed to PoolEntry

fmeheust committed:
Removed changes to PoolEntry

fmeheust committed:
Removed changes to PoolEntry

fmeheust committed:
Removed changes to PoolEntry

fmeheust commented:
Given that this is a Java 11+ version of HikariCP and that request boundaries (Connection.beginRequest() and Connection.endRequest()) are present since Java 9, I have removed the check for JDBC version.
I have added a property that allows to explicitly enable request boundaries. Calls to Connection.beginRequest() and Connection.endRequest() have been moved to HikariPool.getConnection and HikariPool.recycle respectively.

fmeheust commented:
@lfbayer Can you please review my latest changes to this pull request ?

jeandelavarene commented:
@lfbayer could you please review this request? Thank you!

krismohan commented:
Hi @lfbayer any update on the merge? Thanks!

brettwooldridge commented:
Manually resolved conflicts. Merged.
